### DISCUSSION TITLE
Running F' on the Raspberry Pi 5

### DISCUSSION THREAD

**Discussion by kadeangell** [nasa/fprime#2288] [Q&A] [answered]
Hey! I'm just trying to follow the Hello World tutorial on the Raspberry Pi 5, but I get an error when I try to generate the build cache.

This is the output I get:

```shell
(fprime-venv) kade@raspberrypi:~/MyProject $ fprime-util generate      
[INFO] Generating build directory at: /home/kade/MyProject/build-fprime-automatic-native
[INFO] Using toolchain file None for platform default
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Searching for F prime modules in: /home/kade/MyProject/fprime;/home/kade/MyProject
Fatal error: Failed to create the main Isolate. (code 9)
CMake Error at fprime/cmake/required.cmake:25 (message):
   fpp tools not found. Install with:
    'pip install -r "/home/kade/MyProject/fprime/requirements.txt"'
Call Stack (most recent call first):
  fprime/cmake/FPrime.cmake:26 (include)
  CMakeLists.txt:13 (include)


-- Autocoder constants file: /home/kade/MyProject/fprime/config/AcConstants.ini
-- Configuration header directory: /home/kade/MyProject/fprime/config
-- Configuring incomplete, errors occurred!
See also "/home/kade/MyProject/build-fprime-automatic-native/CMakeFiles/CMakeOutput.log".
[ERROR] CMake erred with return code 1. Partial build cache remains. Run purge to clean-up.
```

**Comment by LeStarch** [comment]
This error occurs when you have not installed the python tooling.  

`pip install -r "/home/kade/MyProject/fprime/requirements.txt` *should* fix the issue.

**Reply by kadeangell** [reply]
I tried that already. All the requirements are already satisfied.

**Reply by LeStarch** [reply]
Oh this is quite literally running on the PI, not cross-compiling as we typically do. Try running one of the FPP tools directly:

```
fpp-depend --help
```

This will determine if there is an error or issue with the fpp tools in aarch64.

**Reply by thomas-bc** [reply]
Couple of things I would try to start debugging:
- which version of tools are you using?
  `python --version` `pip freeze | grep fprime`
- Are fpp tools really not on the path?
  `which fpp-to-cpp` -> if nothing, `echo $PATH` to see if the virtual environment got activated properly, and go look in the virtual environment to see if tools got installed properly

**Reply by LeStarch** [reply]
If that works, you can run:

```
fprime-util generate -DFPRIME_SKIP_TOOLS_VERSION_CHECK=ON
```

That might bypass the issue.  Should there be a problem with the FPP tools, however; this will result in more issues.

**Reply by kadeangell** [reply]
Yeah, I'm running directly on the pi. I figured I would try it out, but I'll also try cross-compiling to it later using the workshop-led-blinker tutorial.

I'm using Python 3.11.2, this is the output of `pip freeze | grep fprime`:
```
fprime-fpl-convert-xml==1.0.0
fprime-fpl-extract-xml==1.0.0
fprime-fpl-layout==1.0.0
fprime-fpl-write-pic==1.0.0
fprime-fpp==1.3.0
fprime-gds==3.3.2
fprime-tools==3.3.3
fprime-visual==1.0.0
```

And using `fprime-util generate -DFPRIME_SKIP_TOOLS_VERSION_CHECK=ON` still doesn't work. The error is slightly different; this time it gives me error code 8.

```
(fprime-venv) kade@raspberrypi:~/MyProject $ fprime-util generate -DFPRIME_SKIP_TOOLS_VERSION_CHECK=ON
[INFO] Generating build directory at: /home/kade/MyProject/build-fprime-automatic-native
[INFO] Using toolchain file None for platform default
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Searching for F prime modules in: /home/kade/MyProject/fprime;/home/kade/MyProject
-- Autocoder constants file: /home/kade/MyProject/fprime/config/AcConstants.ini
-- Configuration header directory: /home/kade/MyProject/fprime/config
-- [python3] python3 found at: /home/kade/fprime-venv/bin/python3
-- [fpp-tools] fpp-depend found at: /home/kade/fprime-venv/bin/fpp-depend
-- [fpp-tools] fpp-to-xml found at: /home/kade/fprime-venv/bin/fpp-to-xml
-- [fpp-tools] fpp-to-cpp found at: /home/kade/fprime-venv/bin/fpp-to-cpp
-- [fpp-tools] fpp-locate-defs found at: /home/kade/fprime-venv/bin/fpp-locate-defs
-- Target build toolchain/platform: Linux/Linux
-- Including /home/kade/MyProject/fprime/cmake/platform/Linux.cmake
-- Requiring thread library
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Installation directory: /home/kade/MyProject/build-artifacts
-- Performing CMake source prescan
-- Performing CMake source prescan - DONE
-- Generating FPP location index
Fatal error: Failed to create the main Isolate. (code 8)
CMake Error at fprime/cmake/target/fpp_locs.cmake:54 (message):
  Helper script 'fpp-redirect-helper' exited with reason: 8
Call Stack (most recent call first):
  fprime/cmake/target/fpp_locs.cmake:93 (generate_locations)
  fprime/cmake/target/target.cmake:67 (fpp_locs_add_global_target)
  fprime/cmake/target/target.cmake:67 (cmake_language)
  fprime/cmake/API.cmake:491 (setup_global_target)
  fprime/cmake/FPrime.cmake:68 (register_fprime_target)
  CMakeLists.txt:13 (include)


-- Configuring incomplete, errors occurred!
See also "/home/kade/MyProject/build-fprime-automatic-native/CMakeFiles/CMakeOutput.log".
[ERROR] CMake erred with return code 1. Partial build cache remains. Run purge to clean-up.
```

**Reply by LeStarch** [reply]
Right, before running with a skip, you need to run just an FPP tool and check that it works:

```
fpp-depend --help
```

My guess that will error with: Fatal error: `Failed to create the main Isolate. (code 8)`

**Reply by kadeangell** [reply]
Yep. I got code 9, but essentially the same thing. Is it an issue with how fpp uses Scala? should I try manually installing fpp-tools?

**Reply by LeStarch** [reply]
The issue appears to be with the configuration of our native image build: https://github.com/coursier/coursier/issues/2636.  Likely due to the kernel page size.

Could you tell us the kernel page size on the RPI?  Try: `getconf PAGESIZE`

**Reply by kadeangell** [reply]
16384

**Reply by LeStarch** [reply]
Our build server reports:
```
getconf PAGESIZE
4096
```

A page alignment issue.....this should be reported as an issue to FPP and we can fix it in the next release.

**Reply by kadeangell** [reply]
Okay. Should I go create an issue in the FPP repo? And will this cause an issue if I try and cross-compile using aarch64 and run the compiled binary on the Pi?

**Reply by LeStarch** [reply, marked as answer]
I can make the issue, as I have lots of context to add to it.  Cross-compiling should be unaffected (unless you are running on an `aarch64` cross-compiling host with a kernel page size greater than `4096`.

**Reply by LeStarch** [reply]
You did already!  I'll add context to that one.

**Reply by kadeangell** [reply]
I did! sorry, wasn't sure what to do when you didn't respond for a bit.

**Reply by LeStarch** [reply]
No worries at all, the same result was reached!  Thanks for looking into this!

**Comment by willygett** [comment]
I'm experiencing a related issue while using my newer Macbook which has Apple silicon. After circling back on the same error of "-- [fpp-tools] fpp-depend installed incorrectly" when trying to run the fprime-util generate command, I realized these newer chips run an ARM architecture whereas the F' requires Intel x86. 

I executed the command you gave above and saw this error:
fpp-depend --help
zsh: bad CPU type in executable: fpp-depend

I ended up having to install Rosetta 2 from my terminal to run the x86 code on the ARM architecture. Hope that helps someone who might be experiencing a similar issue.

**Reply by zimri-leisher** [reply]
@bocchino or @Kronos3  might know more about this. Do you support/plan to support ARM executables for the fpp toolchain?

**Reply by LeStarch** [reply]
The current formal release of the FPP tools requires Rosetta 2 on Mac hardware. (aarch64, Intel are supported on Linux).

We have a fix for native mac w/ aarch64  support and expect it to come in the next release.
